smolyay/basis.py

```diff
@@ -403,29 +477,331 @@ def _derivative(self, x):
+            order of derivative. Default is 1.
+
+        Returns
+        -------
+        scalar or ndarray
+            the values of the basis functions.
+        """
+        new_X = self.scale_to_domain(numpy.asarray(X), X_domain)
+        if any(bf._is_complex for bf in self):
+            y = numpy.zeros([len(self)] + list(new_X.shape), dtype="complex_")
+        else:
+            y = numpy.zeros([len(self)] + list(new_X.shape))
+        for i in range(len(self)):
+            y[i, :] = self[i].derivative(new_X, n)
+        y *= ((self.domain[1] - self.domain[0]) / (X_domain[1] - X_domain[0])) ** n
+        return y
+
+
+class ChebyshevFirstKindBasisFunctionSet(BasisFunctionSet):
+    """Set of Chebyshev polynomials of the first kind.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def __init__(self, num_terms):
+        basis_functions = [ChebyshevFirstKind(f) for f in range(num_terms)]
+        super().__init__(basis_functions)
+
+
+class ChebyshevSecondKindBasisFunctionSet(BasisFunctionSet):
+    """Set of Chebyshev polynomials of the second kind.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def __init__(self, num_terms):
+        basis_functions = [ChebyshevSecondKind(f) for f in range(num_terms)]
+        super().__init__(basis_functions)
+
+
+class TrigonometricBasisFunctionSet(BasisFunctionSet):
+    """Set of Trigonmetric equations.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def __init__(self, num_terms):
+        index_trig = numpy.arange(num_terms, dtype=int)
+        frequencies = numpy.where(
+            index_trig % 2 == 1, (1 + index_trig) // 2, -index_trig // 2
+        )
+        basis_functions = [Trigonometric(f) for f in frequencies]
+        super().__init__(basis_functions)
+
+
+class NestedBasisFunctionSet(BasisFunctionSet):
+    """Set of nested basis functions and sample points.
+
+    Parameters
+    ----------
+    basis_functions : list
+        Basis functions in set.
+
+    num_per_level : list
+        number of unique functions per level.
+
+    Raises
+    ------
+    IndexError
+        number of basis function does not match functions in each level.
+    TypeError
+        Basis functions must have the same domain and typing.
+    """
+
+    def __init__(self, basis_functions, num_per_level):
+        self._num_levels = None
+        self._basis_functions = None
+        self._start_level = None
+        self._end_level = None
+        self._num_per_level = None
+
+        # ensure parameters are lists
+        if isinstance(basis_functions, BasisFunction):
+            basis_functions = [basis_functions]
+        if not hasattr(num_per_level, "__len__"):
+            num_per_level = [int(num_per_level)]
+        # check validity of basis functions
+        if sum(num_per_level) != len(basis_functions):
```

Comment:
Maybe you should also move this to after you call the `super()` constructor?
```suggestion
        if numpy.sum(num_per_level) != len(basis_functions):
```